app/src/pm.c

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2023 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/init.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/pm.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#include <zmk/endpoints.h>
+
+#define HAS_WAKERS DT_HAS_COMPAT_STATUS_OKAY(zmk_soft_off_wakeup_sources)
+
+#if HAS_WAKERS
+
+#define DEVICE_WITH_SEP(node_id, prop, idx) DEVICE_DT_GET(DT_PROP_BY_IDX(node_id, prop, idx)),
+
+const struct device *soft_off_wakeup_sources[] = {
+    DT_FOREACH_PROP_ELEM(DT_INST(0, zmk_soft_off_wakeup_sources), wakeup_sources, DEVICE_WITH_SEP)};
+
+#endif
+
+int zmk_pm_soft_off(void) {
+#if IS_ENABLED(CONFIG_PM_DEVICE)
+    size_t device_count;
+    const struct device *devs;
+
+#if !IS_ENABLED(CONFIG_ZMK_SPLIT) || IS_ENABLED(CONFIG_ZMK_SPLIT_ROLE_CENTRAL)
```

Comment:
Probably something for another PR, but there are a lot of these checks for "if split and peripheral" or "if not split or central" in the codebase. Might be useful to create some macros to simplify those.